### PR TITLE
Re-work the profile zone implementation here, and other stuff piling up waiting for CI to pass

### PR DESCRIPTION
**Pull request by tml1024**
* Resolves: # 
* Target version: master 

### TODO

- [ ] ...

### Checklist

- [x] Code is properly formatted
- [x] All commits have Change-Id
- [ ] I have run tests with `make check`
- [x] I have issued `make run` and manually verified that everything looks okay
- [x] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

File: common/ProfileZone.hpp [hunk 1]
```diff
@@ -91,7 +95,7 @@ class ProfileZone
 
     ~ProfileZone()
     {
-        if (m_nCreateTime.time_since_epoch() > std::chrono::time_point<std::chrono::system_clock>().time_since_epoch())
```
Comment (mmeeks) [inline common/ProfileZone.hpp:94]: Wow - we had a system-call for every profile zone destructor?
I really want the ProfileZone stuff to be almost a no-op / zero performance impact when it is not enabled. This is better - but - I'm amazed to see that thing[!] even as it disappears =)

File: kit/Kit.cpp [hunk 2]
```diff
@@ -1799,6 +1803,21 @@ class Document final : public DocumentManagerInterface
...
 
+#if !defined FUZZER && !defined BUILDING_TESTS
+
+// When building the fuzzer we link LOOLWSD.cpp into the same executable so the
+// Protected::addOneRecording() there gets used. When building the unit tests the one in
+// ProfileZone.cpp gets used.
+
+void ProfileZone::addOneRecording(const std::string &sRecording)
+{
+    if (!config::getBool("trace_event[@enable]", false))
```
Comment (mmeeks) [inline kit/Kit.cpp:1814]: I'd expect this to be heinously slow too - for every line we add as we trace; can we have a static / global variable that is set at load time (?)

File: kit/Kit.cpp [hunk 3]
```diff
@@ -1799,6 +1803,21 @@ class Document final : public DocumentManagerInterface
...
+        return;
+
+    singletonDocument->sendTextFrame("trace: \n" + sRecording);
```
Comment (mmeeks) [inline kit/Kit.cpp:1817]: Hmm - we really want to send these one by one? or does the recording contain lots of back to back events? We really want to reduce the performance impact and overhead of performance tracing as much as possible so it doesn't distort what we're trying to measure.

**Review by mmeeks — COMMENTED** [review]
A few queries =)

File: common/ProfileZone.hpp (hunk 1 above)
Comment (tml1024) [reply common/ProfileZone.hpp:94]: Not sure I see any system call there. And anyway, "we had" is an odd expression to use for code that has been there for maybe a week, and has not been used at all.

File: kit/Kit.cpp (hunk 2 above)
Comment (tml1024) [reply kit/Kit.cpp:1814]: Ask @Ashod , who explicitly said I should use that instead of a global variable.

File: kit/Kit.cpp (hunk 3 above)
Comment (tml1024) [reply kit/Kit.cpp:1817]: Yeah, clearly we need to optimise this before even seeing whether it has any impact. Will get right on it.

File: kit/Kit.cpp (hunk 2 above)
Comment (tml1024) [reply kit/Kit.cpp:1814]: You use a global variables. The expert say no, we have an API for that. You use that API. Another expert says no, use a global variable. Repeat.

Comment (tml1024) [reply kit/Kit.cpp:1814]: From Mattermost: "@tml: fyi, the kit now has full access to loolwsd.xml :slightly_smiling_face: so we don't need to pass any new configs to kit explicitly and we don't need any new global or other variables in kit... see ConfigUtil.{hpp,cpp}."

File: common/ProfileZone.hpp (hunk 1 above)
Comment (mmeeks) [reply common/ProfileZone.hpp:94]: You're quite right; I had assumed one of these would have done a gettimeofday or equivalent; but I'm wrong =)

File: kit/Kit.cpp (hunk 2 above)
Comment (mmeeks) [reply kit/Kit.cpp:1841]: heh - so a locally scoped static bool = would make both happy I think.

File: wsd/Storage.cpp [hunk 4]
```diff
@@ -651,6 +652,8 @@ std::unique_ptr<WopiStorage::WOPIFileInfo> WopiStorage::getWOPIFileInfo(const Au
                                                                         const std::string& cookies,
                                                                         LockContext& lockCtx)
 {
+    ProfileZone profileZone("getWOPIFileInfo", { {"url", _fileUrl} });
```
Comment (mmeeks) [inline wsd/Storage.cpp:655]: Lovely =) Ash'll need to convert it to an async pair as/when we get async save in I guess but that's fine.

File: kit/Kit.cpp [hunk 5]
```diff
@@ -1809,13 +1814,46 @@ class Document final : public DocumentManagerInterface
...
+        flushProfileZoneRecordings();
+    else if (profileZoneRecordings.size() == 0 && profileZoneRecordings.capacity() < profileZoneRecordingsCapacity)
+        profileZoneRecordings.reserve(profileZoneRecordingsCapacity);
+    profileZoneRecordings.emplace_back(recording);
 }
+
+#else
+
+static void flushProfileZoneRecordings()
+{
```
Comment (mmeeks) [inline kit/Kit.cpp:1854]: Looks great thanks for that; do we send the frame here?

File: common/Log.cpp [hunk 6]
```diff
@@ -65,9 +67,28 @@ namespace Log
 
         const std::string& getName() const { return _name; }
 
+        void setLevel(const std::string& logLevel) { _logLevel = logLevel; }
+
+        const std::string& getLevel() const { return _logLevel; }
+
         void setLogger(Poco::Logger* logger) { _logger = logger; };
```
Comment (mmeeks) [inline common/Log.cpp:74]: How do we avoid leaking that ? Presumably at the end of the DocumentBroker::poll - we could check if this is set, and unset / free / clean it up (?) I guess it's only for debugging but still =)

Anyhow - looks lovely; thanks Tor!

Comment (tml1024) [reply common/Log.cpp:74]: There is a FIXME about that: 
 ```
            // FIXME: What to do with the previous thread-local logger, if any? Will deleting it
            // destroy also its channel? That won't be good as we use the same channel for all
            // loggers. Best to just leak it?
```

**Pull request merged**